server/src/main/java/org/elasticsearch/bootstrap/Security.java

```diff
@@ -203,21 +210,39 @@ private static Map<String, Set<URL>> readSecuredFiles(
         Map<URL, Policy> pluginPolicies
     ) throws IOException {
         Map<String, Set<URL>> securedFiles = new HashMap<>();
+        Map<String, Set<URL>> securedSettingKeys = new HashMap<>();
 
         for (URL url : mainCodebases) {
-            PolicyUtil.getPolicyPermissions(url, template, environment.tmpFile())
-                .stream()
-                .flatMap(Security::extractSecuredFileName)
-                .map(environment.configFile()::resolve)
-                .forEach(f -> securedFiles.computeIfAbsent(f.toString(), k -> new HashSet<>()).add(url));
+            for (Permission p : PolicyUtil.getPolicyPermissions(url, template, environment.tmpFile())) {
+                readSecuredPermissions(environment, url, p, securedFiles, securedSettingKeys);
+            }
         }
 
         for (var pp : pluginPolicies.entrySet()) {
-            PolicyUtil.getPolicyPermissions(pp.getKey(), pp.getValue(), environment.tmpFile())
-                .stream()
-                .flatMap(Security::extractSecuredFileName)
-                .map(environment.configFile()::resolve)
-                .forEach(f -> securedFiles.computeIfAbsent(f.toString(), k -> new HashSet<>()).add(pp.getKey()));
+            for (Permission p : PolicyUtil.getPolicyPermissions(pp.getKey(), pp.getValue(), environment.tmpFile())) {
+                readSecuredPermissions(environment, pp.getKey(), p, securedFiles, securedSettingKeys);
+            }
+        }
+
+        // compile a Pattern for each setting key we'll be looking for
+        // the key could include a * wildcard
+        List<Map.Entry<Pattern, Set<URL>>> settingPatterns = securedSettingKeys.entrySet()
+            .stream()
+            .map(e -> Map.entry(Pattern.compile(e.getKey()), e.getValue()))
+            .toList();
+
+        for (String setting : environment.settings().keySet()) {
+            for (Map.Entry<Pattern, Set<URL>> ps : settingPatterns) {
+                if (ps.getKey().matcher(setting).matches()) {
+                    // add the setting value to the secured files for these codebase URLs
+                    Path file = environment.configFile().resolve(environment.settings().get(setting));
```

Comment:
We should make sure the file is actually rooted in the config directory, ie a relative path wasn't specified that traverses outside it

Reply:
I've added a subpath check